File: packages/typescript/src/catalogs/banxico/salarios-minimos.ts
/**
 * Salarios Mínimos Catalog
 *
 * Provides access to minimum wage data from Banco de México.
 */

import salariosData from '../../../../shared-data/banxico/salarios_minimos.json';

export interface SalarioMinimoRecord {
  fecha: string;
  salario_minimo: number;
  tipo: string;
  zona: string;
  periodo: string;
  serie: string;
  año: number;
  mes: number;
  fuente: string;
  base_year?: number;
}

export class SalariosMinimosCatalog {
  private static data: SalarioMinimoRecord[] = salariosData as SalarioMinimoRecord[];

  static getData(): SalarioMinimoRecord[] {
    return [...this.data];
  }

  static getPorFechaZona(fecha: string, zona: string = 'general'): SalarioMinimoRecord | null {
    return this.data.find((r) => r.fecha === fecha && r.zona === zona) ?? null;
  }

  static getActualZona(zona: string = 'general'): SalarioMinimoRecord | null {
    const zoneRecords = this.data.filter((r) => r.zona === zona);
    if (zoneRecords.length === 0) return null;

    return zoneRecords.reduce((latest, record) => (record.fecha > latest.fecha ? record : latest));
  }

  static getActualGeneral(): SalarioMinimoRecord | null {
    return this.getActualZona('general');
  }

  static getActualFrontera(): SalarioMinimoRecord | null {
    return this.getActualZona('frontera_norte');
  }

  static getSalarioActualZona(zona: string = 'general'): number | null {
    return this.getActualZona(zona)?.salario_minimo ?? null;
  }
}

export function getSalarioMinimoActualGeneral(): SalarioMinimoRecord | null {
  return SalariosMinimosCatalog.getActualGeneral();
}

export function getSalarioMinimoActualFrontera(): SalarioMinimoRecord | null {
  return SalariosMinimosCatalog.getActualFrontera();
}

File: packages/dart/lib/src/catalogs/banxico/salarios_minimos.dart
/// Salarios Mínimos Catalog
///
/// Catalog of minimum wage data from Banco de México
library;

/// Minimum Wage Catalog
class SalariosMinimosCatalog {
  static List<Map<String, dynamic>>? _data;

  static void _loadData() {
    if (_data != null) return;
    // TODO: Load from packages/shared-data/banxico/salarios_minimos.json
    _data = [];
  }

  static List<Map<String, dynamic>> getData() {
    _loadData();
    return List.from(_data!);
  }

  static Map<String, dynamic>? getActualGeneral() {
    _loadData();
    if (_data!.isEmpty) return null;

    final generalRecords = _data!.where((r) => r['zona'] == 'general').toList();
    if (generalRecords.isEmpty) return null;

    generalRecords
        .sort((a, b) => (b['fecha'] as String).compareTo(a['fecha'] as String));
    return generalRecords.first;
  }

  static Map<String, dynamic>? getActualFrontera() {
    _loadData();
    if (_data!.isEmpty) return null;

    final fronteraRecords =
        _data!.where((r) => r['zona'] == 'frontera_norte').toList();
    if (fronteraRecords.isEmpty) return null;

    fronteraRecords
        .sort((a, b) => (b['fecha'] as String).compareTo(a['fecha'] as String));
    return fronteraRecords.first;
  }

  static double? getSalarioActualZona(String zona) {
    final record =
        zona == 'frontera_norte' ? getActualFrontera() : getActualGeneral();
    return record?['salario_minimo'] as double?;
  }
}
